keps/sig-storage/20200709-pod-pvc-qos-limits.md

```diff
@@ -0,0 +1,255 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+## Summary
+
+Better resource management for iops or bps of block device in pvc.
+In some senarios, we need to limit PV's iops limit or Pod iops on a device or volume at runtime.
+
+For container runtime, dockerd provides blkio related params in docker run to limit a device's iops and bps. [docker reference #block-io-bandwidth-blkio-constraint](https://docs.docker.com/engine/reference/run/#block-io-bandwidth-blkio-constraint)
+
+
+## Motivation
+The proposal would benifit in below two senarios.
+
+* Local Storage(local disk devices), better speed:
+For instances, I want to use local storage(local device) to gain the storage local speed.
+These PVCs in my cluster are blocking devices.
+
+* rootfs (preventing being overused by some pods)
+Besides, the rootfs is shared by all pods in your node, the limitation will help you make node more stable. When one pod in your node heavily uses rootfs, it may effect other pods on the same node. The limitation would help as well.
+
+
+### Goals
+
+- The limit is runtime limitation for block device when it is mounted to the pod. 
+- Only blocking device will be limited with specified volume device id. 
+- This should be implemented with cgroup, so it will only work beyond cgroup capability.
+- An extended feature that can be provided later is that we can limit iops of rootfs for container runtime.  This will reduce the influcences between pods， running on the same host.
+
+
+### Non-Goals
+
+- The limit is runtime limitation for block device when it is mounted to the pod. The limit is not a volume limitation in IaaS's aspect. If the device is used by multi pods, each pod should limit the iops by itself. 
+- For volume capability of iops, it is a physical limitation on device(PV), and this is not the same as the QoS of PVC in this proposal.
+- As cgroup implement has its limitations, we will not mention kernel buffered writings issues with cgroup limitations. This should be fixed or optimized in kernel side. Detailes will be mentioned in the `risks and limitations` below.
+
+
+## Proposal
+
+After open a feature gate "VolumeQos=True" or "DeviceIOPS=true", pod volume qos can be added by annotation like below. The alpha MVP would use annotations.
+
+
+```
+annotations:
+    qos.volume.storage.daocloud.io: >-
+        {"pvc": "snap-03", "iops": {"read": 2000, "write": 1000}, "bps":
```

Comment:
If this is to be volume-based, why not use the pod volume name rather than the PVC? That would incidentally allow for rootfs throttling.

This might not be effective for all kinds of volumes depending on how they're mounted, I suppose, but since the throttling has to be best-effort anyway that might not be a problem in practice.